File: src/main/java/com/gtladd/gtladditions/data/Recipes/NewMachineRecipe/BiologicalSimulation.java
package com.gtladd.gtladditions.data.Recipes.NewMachineRecipe;

import org.gtlcore.gtlcore.common.data.GTLMaterials;
import org.gtlcore.gtlcore.common.data.GTLRecipeTypes;

import com.gregtechceu.gtceu.api.GTValues;
import com.gregtechceu.gtceu.api.data.chemical.material.Material;
import com.gregtechceu.gtceu.common.data.GTMaterials;

import net.minecraft.data.recipes.FinishedRecipe;

import com.gtladd.gtladditions.api.Recipe.GTLAddRecipesTypes;
import com.gtladd.gtladditions.api.Registry.GTLAddRecipeBuilder;

import java.util.Objects;
import java.util.function.Consumer;

public class BiologicalSimulation {

    public BiologicalSimulation() {}

    public static void init(Consumer<FinishedRecipe> provider) {
        Sword[] swords = {
                new Sword("minecraft:diamond_sword", 15, 1),
                new Sword("minecraft:netherite_sword", 5, 5),
                new Sword("avaritia:infinity_sword", 0, 20)
        };

        ItemStacks[] itemStacks = {
                new ItemStacks("blaze", "nether", GTMaterials.Biomass, GTMaterials.SterileGrowthMedium, "minecraft:blaze_rod", 5, 7),
                new ItemStacks("chicken", "overworld", GTMaterials.Biomass, "minecraft:chicken", 75, "minecraft:feather", 40, 2),
                new ItemStacks("cow", "overworld", GTMaterials.Biomass, "minecraft:beef", 75, "minecraft:leather", 25, 2),
                new ItemStacks("drowned", "overworld", GTMaterials.Biomass, "minecraft:rotten_flesh", 75, "minecraft:copper_ingot", 6, 3),
                new ItemStacks("enderman", "end", GTMaterials.Biomass, GTMaterials.SterileGrowthMedium, "minecraft:ender_pearl", 5, 6),
                new ItemStacks("ghast", "nether", GTMaterials.Biomass, GTMaterials.SterileGrowthMedium, "minecraft:gunpowder", 60, "minecraft:ghast_tear", 6, 7),
                new ItemStacks("creeper", "overworld", GTMaterials.Biomass, "minecraft:gunpowder", 80, 2),
                new ItemStacks("zombie", "overworld", GTMaterials.Biomass, "minecraft:rotten_flesh", 75, "minecraft:iron_ingot", 6, "minecraft:carrot", 15, "minecraft:potato", 15, 2),
                new ItemStacks("zombie_villager", "overworld", GTMaterials.Biomass, "minecraft:rotten_flesh", 75, "minecraft:iron_ingot", 6, "minecraft:carrot", 15, "minecraft:potato", 15, 2),
                new ItemStacks("husk", "overworld", GTMaterials.Biomass, "minecraft:rotten_flesh", 75, "minecraft:iron_ingot", 6, "minecraft:carrot", 15, "minecraft:potato", 15, 2),
                new ItemStacks("zombified_piglin", "nether", GTMaterials.Biomass, "minecraft:rotten_flesh", 75, "minecraft:gold_ingot", 6, "minecraft:gold_nugget", 10, 3),
                new ItemStacks("pig", "overworld", GTMaterials.Biomass, "minecraft:porkchop", 80, 2),
                new ItemStacks("sheep", "overworld", GTMaterials.Biomass, "minecraft:mutton", 80, "minecraft:white_wool", 50, 2),
                new ItemStacks("skeleton", "overworld", GTMaterials.Biomass, "minecraft:bone", 75, "minecraft:arrow", 65, 2),
                new ItemStacks("slime", "overworld", GTMaterials.Biomass, "minecraft:slime_ball", 50, 3),
                new ItemStacks("spider", "overworld", GTMaterials.Biomass, "minecraft:string", 70, "minecraft:spider_eye", 20, 2),
                new ItemStacks("vindicator", "overworld", GTMaterials.Biomass, "minecraft:emerald", 10, 3),
                new ItemStacks("witch", "overworld", GTMaterials.Biomass, "minecraft:stick", 50, "minecraft:gunpowder", 35, "minecraft:sugar", 35, "minecraft:glass_bottle", 35, "minecraft:redstone", 6, "minecraft:glowstone_dust", 6, "minecraft:spider_eye", 6, 3),
                new ItemStacks("wither_skeleton", "nether", GTMaterials.Biomass, GTLMaterials.BiohmediumSterilized, "minecraft:bone", 75, "minecraft:coal", 65, "minecraft:wither_skeleton_skull", 5, 8),
                new ItemStacks("rabbit", "overworld", GTMaterials.Biomass, "minecraft:rabbit", 70, "minecraft:rabbit_hide", 10, "minecraft:rabbit_foot", 5, 3),
                new ItemStacks("donkey", "overworld", GTMaterials.Biomass, "minecraft:leather", 50, 2),
                new ItemStacks("llama", "overworld", GTMaterials.Biomass, "minecraft:leather", 50, 2),
                new ItemStacks("cat", "overworld", GTMaterials.Biomass, "minecraft:string", 50, 2),
                new ItemStacks("panda", "overworld", GTMaterials.Biomass, "minecraft:bamboo", 50, 3),
                new ItemStacks("polar_bear", "overworld", GTMaterials.Biomass, "minecraft:cod", 50, "minecraft:salmon", 50, 3)

        };

        for (ItemStacks i : itemStacks) {
            for (Sword s : swords) {
                generateRecipe(i, s, provider);
            }
            setspawneggreicpes(i, provider);
        }
        generateSpecialRecipes(provider);
    }

    private static void generateRecipe(ItemStacks item, Sword sword, Consumer<FinishedRecipe> provider) {
        GTLAddRecipeBuilder builder = new GTLAddRecipeBuilder(item.name + (sword.damage > 10 ? "_1" : (sword.damage > 0 ? "_2" : "_3")), GTLAddRecipesTypes.BIOLOGICAL_SIMULATION)
                .notConsumable("minecraft:" + item.name + "_spawn_egg")
                .notConsumable("kubejs:" + item.data + "_data");

        if (Objects.equals(sword.name, "avaritia:infinity_sword")) {
            builder.notConsumable(sword.name);
        } else {
            builder.chancedInputItems(sword.name, sword.damage, 0);
        }

        builder.inputFluids(item.F1.getFluid(1000 / sword.factor));

        if (item.F2 != null) {
            builder.inputFluids(item.F2.getFluid(500 / sword.factor));
        }

        addOutputItems(builder, item);
        builder.TierEUtVA(item.EUt).duration(400 / sword.factor).save(provider);
    }

    private static void setspawneggreicpes(ItemStacks item, Consumer<FinishedRecipe> provider) {
        GTLAddRecipeBuilder builder = new GTLAddRecipeBuilder(item.name + "_spawn_egg", GTLRecipeTypes.INCUBATOR_RECIPES)
                .inputItems("minecraft:bone", 4)
                .inputFluids(GTMaterials.Biomass.getFluid(1000))
                .inputFluids(GTMaterials.Milk.getFluid(1000));
        addInputItems(builder, item);
        builder.outputItems("minecraft:" + item.name + "_spawn_egg").TierEUtVA(3).duration(1200).save(provider);
    }

    private static void addOutputItems(GTLAddRecipeBuilder builder, ItemStacks item) {
        builder.chancedOutputItems(item.O1, item.O1f, 0);
        if (item.O2 != null) builder.chancedOutputItems(item.O2, item.O2f, 0);
        if (item.O3 != null) builder.chancedOutputItems(item.O3, item.O3f, 0);
        if (item.O4 != null) builder.chancedOutputItems(item.O4, item.O4f, 0);
        if (item.O5 != null) builder.chancedOutputItems(item.O5, item.O5f, 0);
        if (item.O6 != null) builder.chancedOutputItems(item.O6, item.O6f, 0);
        if (item.O7 != null) builder.chancedOutputItems(item.O7, item.O7f, 0);
    }

    private static void addInputItems(GTLAddRecipeBuilder builder, ItemStacks item) {
        if (item.name.equals("witch")) {
            builder.inputItems("minecraft:redstone", 4).inputItems("minecraft:glowstone_dust", 4)
                    .inputItems("minecraft:sugar", 4).inputItems("minecraft:glass_bottle", 4);
            return;
        }
        if (!Objects.equals(item.O1, "minecraft:bone")) builder.inputItems(item.O1, 4);
        if (item.O2 != null) builder.inputItems(item.O2, 4);
        if (item.O3 != null) builder.inputItems(item.O3, 4);
        if (item.O4 != null) builder.inputItems(item.O4, 4);
    }

    private static void generateSpecialRecipes(Consumer<FinishedRecipe> provider) {
        new GTLAddRecipeBuilder("nether_star", GTLAddRecipesTypes.BIOLOGICAL_SIMULATION)
                .notConsumable("gtceu:nether_star_block")
                .notConsumable("kubejs:nether_data", 64)
                .notConsumable("avaritia:infinity_sword")
                .inputFluids(GTMaterials.Biomass.getFluid(50))
                .inputFluids(GTLMaterials.BiohmediumSterilized.getFluid(50))
                .chancedOutputItems("minecraft:nether_star", 15, 0)
                .duration(100)
                .EUt(GTValues.VA[GTValues.UV])
                .save(provider);

        new GTLAddRecipeBuilder("dragon_egg", GTLAddRecipesTypes.BIOLOGICAL_SIMULATION)
                .notConsumable("minecraft:dragon_head")
                .notConsumable("kubejs:end_data", 64)
                .notConsumable("avaritia:infinity_sword")
                .inputFluids(GTMaterials.Biomass.getFluid(50))
                .inputFluids(GTLMaterials.BiohmediumSterilized.getFluid(50))
                .outputItems("minecraft:dragon_egg")
                .duration(100)
                .EUt(GTValues.VA[GTValues.UV])
                .save(provider);
    }

    static class ItemStacks {

        String name;
        String data;
        Material F1;
        Material F2;
        String O1;
        double O1f;
        String O2;
        double O2f;
        String O3;
        double O3f;
        String O4;
        double O4f;
        String O5;
        double O5f;
        String O6;
        double O6f;
        String O7;
        double O7f;
        int EUt;

        public ItemStacks(String name, String data, Material f1, String o1, double o1f, String o2, double o2f, String o3, double o3f, String o4, double o4f, String o5, double o5f, String o6, double o6f, String o7, double o7f, int EUt) {
            this.name = name;
            this.data = data;
            F1 = f1;
            O1 = o1;
            O1f = o1f;
            O2 = o2;
            O2f = o2f;
            O3 = o3;
            O3f = o3f;
            O4 = o4;
            O4f = o4f;
            O5 = o5;
            O5f = o5f;
            O6 = o6;
            O6f = o6f;
            O7 = o7;
            O7f = o7f;
            this.EUt = EUt;
        }

        public ItemStacks(String name, String data, Material f1, String o1, double o1f, int EUt) {
            this.name = name;
            this.data = data;
            F1 = f1;
            O1 = o1;
            O1f = o1f;
            this.EUt = EUt;
        }

        public ItemStacks(String name, String data, Material f1, Material f2, String o1, double o1f, int EUt) {
            this.name = name;
            this.data = data;
            F1 = f1;
            F2 = f2;
            O1 = o1;
            O1f = o1f;
            this.EUt = EUt;
        }

        public ItemStacks(String name, String data, Material f1, String o1, double o1f, String o2, double o2f, int EUt) {
            this.name = name;
            this.data = data;
            F1 = f1;
            O1 = o1;
            O1f = o1f;
            O2 = o2;
            O2f = o2f;
            this.EUt = EUt;
        }

        public ItemStacks(String name, String data, Material f1, String o1, double o1f, String o2, double o2f, String o3, double o3f, int EUt) {
            this.name = name;
            this.data = data;
            F1 = f1;
            O1 = o1;
            O1f = o1f;
            O2 = o2;
            O2f = o2f;
            O3 = o3;
            O3f = o3f;
            this.EUt = EUt;
        }

        public ItemStacks(String name, String data, Material f1, String o1, double o1f, String o2, double o2f, String o3, double o3f, String o4, double o4f, int EUt) {
            this.name = name;
            this.data = data;
            F1 = f1;
            O1 = o1;
            O1f = o1f;
            O2 = o2;
            O2f = o2f;
            O3 = o3;
            O3f = o3f;
            O4 = o4;
            O4f = o4f;
            this.EUt = EUt;
        }

        public ItemStacks(String name, String data, Material f1, Material f2, String o1, double o1f, String o2, double o2f, int EUt) {
            this.name = name;
            this.data = data;
            F1 = f1;
            F2 = f2;
            O1 = o1;
            O1f = o1f;
            O2 = o2;
            O2f = o2f;
            this.EUt = EUt;
        }

        public ItemStacks(String name, String data, Material f1, Material f2, String o1, double o1f, String o2, double o2f, String o3, double o3f, int EUt) {
            this.name = name;
            this.data = data;
            F1 = f1;
            F2 = f2;
            O1 = o1;
            O1f = o1f;
            O2 = o2;
            O2f = o2f;
            O3 = o3;
            O3f = o3f;
            this.EUt = EUt;
        }
    }

    static class Sword {

        String name;
        double damage;
        int factor;

        public Sword(String name, double damage, int factor) {
            this.name = name;
            this.damage = damage;
            this.factor = factor;
        }
    }
}

File: src/main/java/com/gtladd/gtladditions/data/recipes/newmachinerecipe/BiologicalSimulation.kt
package com.gtladd.gtladditions.data.recipes.newmachinerecipe

import com.gregtechceu.gtceu.api.GTValues.UV
import com.gregtechceu.gtceu.api.GTValues.VA
import com.gregtechceu.gtceu.common.data.GTMaterials.Biomass
import com.gregtechceu.gtceu.common.data.GTMaterials.Milk
import com.gregtechceu.gtceu.data.recipe.builder.GTRecipeBuilder
import com.gtladd.gtladditions.GTLAdditions.Companion.id
import com.gtladd.gtladditions.common.recipe.GTLAddRecipesTypes
import net.minecraft.data.recipes.FinishedRecipe
import org.gtlcore.gtlcore.common.data.GTLMaterials.BiohmediumSterilized
import org.gtlcore.gtlcore.common.data.GTLRecipeTypes.INCUBATOR_RECIPES
import org.gtlcore.gtlcore.utils.Registries.getItemStack
import java.util.function.Consumer

object BiologicalSimulation {
    fun init(provider: Consumer<FinishedRecipe?>) {
        val swords = listOf(
            Sword("minecraft:diamond_sword", 15, 1),
            Sword("minecraft:netherite_sword", 5, 5),
            Sword("avaritia:infinity_sword", 0, 20)
        )
        val biologicals = listOf(
            Biological("blaze", "nether", "minecraft:blaze_rod", 500, 7),
            Biological("chicken", "overworld", "minecraft:chicken", 7500, "minecraft:feather", 4000, "minecraft:egg", 1000, 2),
            Biological("cow", "overworld", "minecraft:beef", 7500, "minecraft:leather", 2500, 2),
            Biological("drowned", "overworld", "minecraft:rotten_flesh", 7500, "minecraft:copper_ingot", 600, 3),
            Biological("enderman", "end", "minecraft:ender_pearl", 500, 6),
            Biological("ghast", "nether", "minecraft:gunpowder", 6000, "minecraft:ghast_tear", 600, 7),
            Biological("creeper", "overworld", "minecraft:gunpowder", 8000, 2),
            Biological("zombie", "overworld", "minecraft:rotten_flesh", 7500, "minecraft:iron_ingot", 600, "minecraft:carrot", 1500, "minecraft:potato", 1500, 2),
            Biological("zombie_villager", "overworld", "minecraft:rotten_flesh", 7500, "minecraft:iron_ingot", 600, "minecraft:carrot", 1500, "minecraft:potato", 1500, 2),
            Biological("husk", "overworld", "minecraft:rotten_flesh", 7500, "minecraft:iron_ingot", 600, "minecraft:carrot", 1500, "minecraft:potato", 1500, 2),
            Biological("zombified_piglin", "nether", "minecraft:rotten_flesh", 7500, "minecraft:gold_ingot", 600, "minecraft:gold_nugget", 1000, 3),
            Biological("pig", "overworld", "minecraft:porkchop", 8000, 2),
            Biological("sheep", "overworld", "minecraft:mutton", 8000, "minecraft:white_wool", 5000, 2),
            Biological("skeleton", "overworld", "minecraft:bone", 7500, "minecraft:arrow", 6500, 2),
            Biological("slime", "overworld", "minecraft:slime_ball", 5000, 3),
            Biological("spider", "overworld", "minecraft:string", 7000, "minecraft:spider_eye", 2000, 2),
            Biological("vindicator", "overworld", "minecraft:emerald", 1000, 3),
            Biological("witch", "overworld", "minecraft:stick", 5000, "minecraft:gunpowder", 3500, "minecraft:sugar", 3500, "minecraft:glass_bottle", 3500, "minecraft:redstone", 600, "minecraft:glowstone_dust", 600, "minecraft:spider_eye", 600, 3),
            Biological("wither_skeleton", "nether", "minecraft:bone", 7500, "minecraft:coal", 6500, "minecraft:wither_skeleton_skull", 500, 8),
            Biological("rabbit", "overworld", "minecraft:rabbit", 7000, "minecraft:rabbit_hide", 1000, "minecraft:rabbit_foot", 500, 3),
            Biological("donkey", "overworld", "minecraft:leather", 5000, 2),
            Biological("llama", "overworld", "minecraft:leather", 5000, 2),
            Biological("cat", "overworld", "minecraft:string", 5000, 2),
            Biological("panda", "overworld", "minecraft:bamboo", 5000, 3),
            Biological("polar_bear", "overworld", "minecraft:cod", 5000, "minecraft:salmon", 5000, 3)
        )
        for (biological in biologicals) {
            for (sword in swords) {
                generateRecipe(biological, sword, provider)
            }
            setSpawnEggRecipes(biological, provider)
        }
        generateSpecialRecipes(provider)
    }

    private fun generateRecipe(item: Biological, sword: Sword, provider: Consumer<FinishedRecipe?>) {
        val builder = GTLAddRecipesTypes.BIOLOGICAL_SIMULATION.recipeBuilder(
            id(item.name + (if (sword.damage > 10) "_1" else (if (sword.damage > 0) "_2" else "_3")))
        ).notConsumable(getItemStack("minecraft:" + item.name + "_spawn_egg"))
            .notConsumable(getItemStack("kubejs:" + item.data + "_data"))
        if (sword.name == "avaritia:infinity_sword") builder.notConsumable(getItemStack(sword.name))
        else builder.chancedInput(getItemStack(sword.name), sword.damage, 0)
        builder.inputFluids(Biomass.getFluid((1000 / sword.factor).toLong()))
        addOutputItems(builder, item, sword)
        builder.EUt(VA[item.eut].toLong()).duration(400 / sword.factor).save(provider)
    }

    private fun setSpawnEggRecipes(item: Biological, provider: Consumer<FinishedRecipe?>) {
        if (item.name == "cow") return
        val builder = INCUBATOR_RECIPES.recipeBuilder(id(item.name + "_spawn_egg"))
            .inputItems(getItemStack("minecraft:bone", 4))
            .inputFluids(Biomass.getFluid(1000))
            .inputFluids(Milk.getFluid(1000))
        addInputItems(builder, item)
        builder.outputItems(getItemStack("minecraft:" + item.name + "_spawn_egg"))
            .EUt(VA[3].toLong()).duration(1200).save(provider)
    }

    private fun addOutputItems(builder: GTRecipeBuilder, item: Biological, sword: Sword) {
        val outputs = listOf(
            Triple(item.o1, item.o1f, true),
            Triple(item.o2, item.o2f, item.o2 != null),
            Triple(item.o3, item.o3f, item.o3 != null),
            Triple(item.o4, item.o4f, item.o4 != null),
            Triple(item.o5, item.o5f, item.o5 != null),
            Triple(item.o6, item.o6f, item.o6 != null),
            Triple(item.o7, item.o7f, item.o7 != null)
        )

        for ((outputItem, chance, shouldAdd) in outputs) {
            if (shouldAdd && outputItem != null) {
                builder.chancedOutput(getItemStack(outputItem, sword.factor * 2), chance, 0)
            }
        }
    }

    private fun addInputItems(builder: GTRecipeBuilder, item: Biological) {
        if (item.name == "witch") {
            builder.inputItems(getItemStack("minecraft:redstone", 4))
                .inputItems(getItemStack("minecraft:glowstone_dust", 4))
                .inputItems(getItemStack("minecraft:sugar", 4))
                .inputItems(getItemStack("minecraft:glass_bottle", 4))
            return
        }

        val inputs = listOfNotNull(
            item.o1.takeIf { it != "minecraft:bone" },
            item.o2,
            item.o3,
            item.o4
        )

        for (input in inputs) {
            builder.inputItems(getItemStack(input, 4))
        }

        when (item.name) {
            "cat", "zombie", "donkey", "creeper" -> builder.circuitMeta(1)
            "zombie_villager", "llama" -> builder.circuitMeta(2)
            "husk" -> builder.circuitMeta(3)
        }
    }

    private fun generateSpecialRecipes(provider: Consumer<FinishedRecipe?>) {
        GTLAddRecipesTypes.BIOLOGICAL_SIMULATION.recipeBuilder(id("nether_star"))
            .notConsumable(getItemStack("gtceu:nether_star_block"))
            .notConsumable(getItemStack("kubejs:nether_data", 64))
            .notConsumable(getItemStack("avaritia:infinity_sword"))
            .inputFluids(Biomass.getFluid(50))
            .inputFluids(BiohmediumSterilized.getFluid(50))
            .chancedOutput(getItemStack("minecraft:nether_star"), 1500, 0)
            .duration(100)
            .EUt(VA[UV].toLong())
            .save(provider)

        GTLAddRecipesTypes.BIOLOGICAL_SIMULATION.recipeBuilder(id("dragon_egg"))
            .notConsumable(getItemStack("minecraft:dragon_head"))
            .notConsumable(getItemStack("kubejs:end_data", 64))
            .notConsumable(getItemStack("avaritia:infinity_sword"))
            .inputFluids(Biomass.getFluid(50))
            .inputFluids(BiohmediumSterilized.getFluid(50))
            .outputItems(getItemStack("minecraft:dragon_egg"))
            .duration(100)
            .EUt(VA[UV].toLong())
            .save(provider)
    }

    internal data class Biological(
        val name: String,
        val data: String?,
        val o1: String,
        val o1f: Int,
        val o2: String?,
        val o2f: Int,
        val o3: String?,
        val o3f: Int,
        val o4: String?,
        val o4f: Int,
        val o5: String?,
        val o5f: Int,
        val o6: String?,
        val o6f: Int,
        val o7: String?,
        val o7f: Int,
        val eut: Int
    ) {
        constructor(name: String, data: String?, o1: String, o1f: Int, eut: Int) :
            this(name, data, o1, o1f, null, 0, null, 0, null, 0, null, 0, null, 0, null, 0, eut)

        constructor(name: String, data: String?, o1: String, o1f: Int, o2: String?, o2f: Int, eut: Int) :
            this(name, data, o1, o1f, o2, o2f, null, 0, null, 0, null, 0, null, 0, null, 0, eut)

        constructor(name: String, data: String?, o1: String, o1f: Int, o2: String?, o2f: Int, o3: String?, o3f: Int, eut: Int) :
            this(name, data, o1, o1f, o2, o2f, o3, o3f, null, 0, null, 0, null, 0, null, 0, eut)

        constructor(name: String, data: String?, o1: String, o1f: Int, o2: String?, o2f: Int, o3: String?, o3f: Int, o4: String?, o4f: Int, eut: Int) :
            this(name, data, o1, o1f, o2, o2f, o3, o3f, o4, o4f, null, 0, null, 0, null, 0, eut)
    }

    internal data class Sword(
        val name: String,
        val damage: Int,
        val factor: Int
    )
}
